lib/Dialect/SV/Transforms/HWLegalizeModules.cpp

```diff
@@ -52,52 +59,234 @@ struct HWLegalizeModulesPass
+        auto index = indexOp.getIndex();
+        if (auto *definingOp = index.getDefiningOp())
+          if (isa<hw::ConstantOp>(definingOp))
+            return true;
+
+        // Skip index ops with unpacked arrays.
+        auto inout = indexOp.getInput().getType();
+        if (isa<hw::UnpackedArrayType>(inout.getElementType()))
+          return true;
+
+        // Generate case value element lookups.
+        auto ty = hw::type_cast<hw::ArrayType>(inout.getElementType());
+        OpBuilder builder(&op);
+        SmallVector<Value> caseValues;
+        for (size_t i = 0; i < ty.getNumElements(); i++) {
+          auto loc = op.getLoc();
+          auto index = builder.createOrFold<hw::ConstantOp>(
+              loc, APInt(llvm::Log2_64_Ceil(ty.getNumElements()), i));
+          auto element = builder.create<sv::ArrayIndexInOutOp>(
+              loc, indexOp.getInput(), index);
+          auto readElement = builder.create<sv::ReadInOutOp>(loc, element);
+          caseValues.push_back(readElement);
+        }
+
+        // Transform array index op into casez statement.
+        auto theWire = lowerLookupToCasez(op, indexOp.getInput(), index,
+                                          ty.getElementType(), caseValues);
+
+        // Replace uses and clean up.
+        indexOp.getResult().replaceAllUsesWith(theWire);
+        indexOp->erase();
+        anythingChanged = true;
+        return true;
+      })
+      .Case<sv::PAssignOp>([&](sv::PAssignOp assignOp) {
+        // Skip assignments to non-array types.
+        auto inout = assignOp.getDest().getType();
+        if (!isa<hw::ArrayType>(inout.getElementType()))
+          return true;
 
-  // array_get(idx, array_create(a,b,c,d)) ==> casez(idx).
-  Value index = getOp.getIndex();
+        // Transform assignment into individual assignments for each array
+        // element.
+        auto ty = hw::type_cast<hw::ArrayType>(inout.getElementType());
+        OpBuilder builder(assignOp);
+        for (size_t i = 0; i < ty.getNumElements(); i++) {
+          auto loc = op.getLoc();
+          auto index = builder.createOrFold<hw::ConstantOp>(
+              loc, APInt(llvm::Log2_64_Ceil(ty.getNumElements()), i));
+          auto dstElement = builder.create<sv::ArrayIndexInOutOp>(
+              loc, assignOp.getDest(), index);
+          auto srcElement =
+              builder.create<hw::ArrayGetOp>(loc, assignOp.getSrc(), index);
+          builder.create<sv::PAssignOp>(loc, dstElement, srcElement);
+        }
+
+        // Remove original assignment.
+        assignOp.erase();
+        anythingChanged = true;
+        return true;
+      })
+      .Case<sv::RegOp>([&](sv::RegOp regOp) {
+        // Skip non-array regs.
+        if (!isa<hw::ArrayType>(regOp.getElementType()))
+          return true;
+
+        // Transform reg into individual regs for each array element.
+        auto ty = hw::type_cast<hw::ArrayType>(regOp.getElementType());
```

Comment:
nit: Please dyn_cast and null guard 
```suggestion
        auto ty = hw::type_dyn_cast<hw::ArrayType>(regOp.getElementType()))
        if(!ty) return true;
        // Transform reg into individual regs for each array element.
```